File: sdk/go/pkg/artifact/rsync.go
package artifact

import (
	"fmt"

	api "github.com/ALT-F4-LLC/vorpal/sdk/go/pkg/api/artifact"
	"github.com/ALT-F4-LLC/vorpal/sdk/go/pkg/config"
)

func Rsync(context *config.ConfigContext) (*string, error) {
	name := "rsync"
	version := "3.4.1"

	sourcePath := fmt.Sprintf("https://sdk.vorpal.build/source/rsync-%s.tar.gz", version)
	source := NewArtifactSource(name, sourcePath).Build()

	stepScript := fmt.Sprintf(`mkdir -p "$VORPAL_OUTPUT"
pushd ./source/%s/%s-%s
./configure --prefix="$VORPAL_OUTPUT" --disable-openssl --disable-xxhash --disable-zstd --disable-lz4
make
make install`, name, name, version)

	step, err := Shell(context, []*string{}, []string{}, stepScript, nil)
	if err != nil {
		return nil, err
	}

	return NewArtifact(name, []*api.ArtifactStep{step}, config.SYSTEMS).
		WithAliases([]string{fmt.Sprintf("%s:%s", name, version)}).
		WithSources([]*api.ArtifactSource{&source}).
		Build(context)
}

File: sdk/python/src/vorpal_sdk/artifact/rsync.py
from __future__ import annotations

from typing import TYPE_CHECKING

from vorpal_sdk.artifact import Artifact, ArtifactSource
from vorpal_sdk.step import shell
from vorpal_sdk.system import SYSTEMS

if TYPE_CHECKING:
    from vorpal_sdk.context import ConfigContext


class Rsync:
    """Builder for the rsync artifact."""

    def build(self, context: ConfigContext) -> str:
        name = "rsync"
        version = "3.4.1"

        path = f"https://sdk.vorpal.build/source/rsync-{version}.tar.gz"
        source = ArtifactSource(name, path).build()

        step_script = (
            f'mkdir -p "$VORPAL_OUTPUT"\n'
            f'pushd ./source/{name}/{name}-{version}\n'
            f'./configure'
            f' --prefix="$VORPAL_OUTPUT"'
            f' --disable-openssl'
            f' --disable-xxhash'
            f' --disable-zstd'
            f' --disable-lz4\n'
            f'make\n'
            f'make install'
        )

        steps = [shell(context, [], [], step_script, [])]

        return (
            Artifact(name, steps, SYSTEMS)
            .with_aliases([f"{name}:{version}"])
            .with_sources([source])
            .build(context)
        )
